ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -0,0 +1,364 @@
+    BlobContainerClient blobContainerClient;
+    try {
+      blobContainerClient = azureStorageClient.getBlobContainerClient(String.valueOf(partitionId));
+    } catch (BlobStorageException blobStorageException) {
+      if (blobStorageException.getErrorCode().equals(BlobErrorCode.CONTAINER_NOT_FOUND)) {
+        logger.error("Azure blob storage container for partition {} not found due to {}", partitionId,
+            blobStorageException.getServiceMessage());
+        return null;
+      }
+      vcrMetrics.azureStoreContainerGetError.inc();
+      logger.error("Failed to get Azure blob storage container for partition {} due to {}", partitionId,
+          blobStorageException.getServiceMessage());
+      throw blobStorageException;
+    }
+    return blobContainerClient;
+  }
+
+  /**
+   * Creates blob container in Azure for an Ambry partition
+   * @param partitionId Ambry partition
+   * @return blobContainerClient
+   */
+  protected BlobContainerClient createBlobStore(String partitionId) {
+    BlobContainerClient blobContainerClient;
+    try {
+      blobContainerClient = azureStorageClient.createBlobContainer(String.valueOf(partitionId));
+    } catch (BlobStorageException blobStorageException) {
+      if (blobStorageException.getErrorCode().equals(BlobErrorCode.CONTAINER_ALREADY_EXISTS)) {
+        logger.info("Azure blob storage container for partition {} already exists", partitionId);
+        return getBlobStore(partitionId);
+      }
+      vcrMetrics.azureStoreContainerGetError.inc();
+      logger.error("Failed to create Azure blob storage container for partition {} due to {}", partitionId,
+          blobStorageException.getServiceMessage());
+      throw blobStorageException;
+    }
+    return blobContainerClient;
+  }
+
+  /**
+   * Creates an object that stores blobs in Azure blob storage
+   * @param partitionId Partition ID
+   * @return {@link CloudBlobStoreV2}
+   */
+  protected BlobContainerClient createOrGetBlobStore(String partitionId) {
+    // Get or create container
+    BlobContainerClient blobContainerClient = partitionToAzureStore.get(partitionId);
+    if (blobContainerClient == null) {
+      blobContainerClient = createBlobStore(partitionId);
+    }
+
+    // If it is still null, throw the error and emit a metric
+    if (blobContainerClient == null) {
+      vcrMetrics.azureStoreContainerGetError.inc();
+      String errMsg = String.format("Azure blob storage container for partition %s is null", partitionId);
+      logger.error(errMsg);
+      throw new RuntimeException(errMsg);
+    }
+
+    partitionToAzureStore.put(partitionId, blobContainerClient);
+    return blobContainerClient;
+  }
+
+  @Override
+  public boolean uploadBlob(BlobId blobId, long inputLength, CloudBlobMetadata cloudBlobMetadata,
+      InputStream blobInputStream) throws CloudStorageException {
+    /**
+     * Current impl of this fn is inefficient because the caller does a 0-4MB memcpy from src to dst stream
```

Comment:
I think copying the `stream` to `ByteBuffer` is needed for retries. During retry, we are rewinding the byte buffer and creating a fresh stream from it again. If we pass the original stream directly, it may happen that part of the stream is written to socket before erroring out. In that case, I think we won't have entire content to retry.

Reply:
To get rid of all these quirks, I plan to let azure-sdk take care of retries and not cloud-blob store.